File: src/agent.ts
import { Ajv, type ValidateFunction } from "ajv";
import {
  DEFAULT_PROJECTION,
  toolSchemas,
  type ProjectionSettings,
} from "./context.js";
import type {
  AssistantMessage,
  JsonValue,
  Llm,
  LlmStreamEvent,
  UnifiedRequest,
} from "./protocol.js";
import type { Context } from "./runtime.js";
import {
  buildRequest,
  replayTrace,
  SESSION_LOG,
  type SessionEvent,
  type SessionLog,
  type TraceItem,
} from "./session.js";

export interface RunTurnResult {
  finalMessage: AssistantMessage;
  steps: number;
  turnId: string;
  stepIds: string[];
  trace: TraceItem[];
  requests: UnifiedRequest[];
}

export interface CheckpointProposal {
  summary: string;
  coveredThroughEventId: number;
}

export interface AgentOptions {
  llm: Llm;
  context: Context;
  systemPrompt: string;
  dynamicContext?: (step: number) => string;
  projection?: ProjectionSettings;
  checkpointBeforeStep?: (input: {
    step: number;
    events: readonly SessionEvent[];
  }) => CheckpointProposal | undefined;
  maxSteps?: number;
  onModelEvent?: (event: { stepId: string; event: LlmStreamEvent }) => void;
}

export class Agent {
  readonly #llm: Llm;
  readonly #context: Context;
  readonly #validators = new Map<string, { schema: string; validate: ValidateFunction }>();
  readonly #ajv = new Ajv({ allErrors: true, strict: false });
  readonly #systemPrompt: string;
  readonly #dynamicContext: (step: number) => string;
  readonly #maxSteps: number;
  readonly #projection: ProjectionSettings;
  readonly #session: SessionLog;
  readonly #checkpointBeforeStep?: AgentOptions["checkpointBeforeStep"];
  readonly #onModelEvent?: AgentOptions["onModelEvent"];

  constructor(options: AgentOptions) {
    this.#llm = options.llm;
    this.#context = options.context;
    this.#systemPrompt = options.systemPrompt;
    this.#dynamicContext = options.dynamicContext ?? ((step) => `Current step: ${step}`);
    this.#maxSteps = options.maxSteps ?? 8;
    this.#projection = options.projection ?? DEFAULT_PROJECTION;
    this.#session = this.#context.use(SESSION_LOG);
    this.#checkpointBeforeStep = options.checkpointBeforeStep;
    this.#onModelEvent = options.onModelEvent;
  }

  async runTurn(userInput: string): Promise<RunTurnResult> {
    const eventStart = this.#session.events.length;
    const turnId = this.#session.nextTurnId();
    this.#session.append({ type: "turn/start", turnId });
    this.#session.append({ type: "user/message", turnId, content: userInput });
    const requests: UnifiedRequest[] = [];
    const stepIds: string[] = [];

    for (let step = 1; step <= this.#maxSteps; step += 1) {
      const checkpoint = this.#checkpointBeforeStep?.({ step, events: this.#session.events });
      if (checkpoint) {
        this.#session.checkpoint(checkpoint.summary, checkpoint.coveredThroughEventId);
      }
      const stepId = `${turnId}-step-${step}`;
      stepIds.push(stepId);
      this.#session.append({ type: "step/start", turnId, stepId, ordinal: step });
      this.#session.append({
        type: "request/header",
        stepId,
        provider: this.#llm.provider,
        model: this.#llm.model,
        system: this.#context.compilePrompt(this.#systemPrompt),
        tools: toolSchemas(this.#context.listTools()),
        dynamicContext: this.#dynamicContext(step),
        projection: this.#projection,
      });
      const request = buildRequest(this.#session.events, stepId);
      requests.push(structuredClone(request));

      let response;
      for await (const event of this.#llm.stream(request)) {
        this.#onModelEvent?.({ stepId, event: structuredClone(event) });
        if (event.type === "response") {
          if (response) throw new Error("LLM stream returned more than one final response.");
          response = event.response;
        }
      }
      if (!response) throw new Error("LLM stream ended without a final response.");
      const assistant = response.message;
      this.#session.append({ type: "assistant/message", stepId, content: assistant.content });
      for (const call of assistant.toolCalls) {
        this.#session.append({ type: "tool/call", stepId, call });
      }

      if (assistant.toolCalls.length === 0) {
        this.#session.append({ type: "step/end", stepId, outcome: "complete" });
        this.#session.append({ type: "turn/end", turnId, outcome: "completed" });
        return {
          finalMessage: assistant,
          steps: step,
          turnId,
          stepIds,
          trace: replayTrace(this.#session.events.slice(eventStart)),
          requests,
        };
      }

      for (const call of assistant.toolCalls) {
        const result = await this.#executeTool(call.name, call.arguments);
        this.#session.append({
          type: "tool/result",
          stepId,
          toolCallId: call.id,
          name: call.name,
          content: JSON.stringify(result),
        });
      }
      this.#session.append({ type: "step/end", stepId, outcome: "tool-calls" });
    }

    this.#session.append({ type: "turn/end", turnId, outcome: "max-steps" });
    throw new Error(`Agent exceeded maxSteps (${this.#maxSteps}) before finishing the turn.`);
  }

  async #executeTool(name: string, input: JsonValue): Promise<JsonValue> {
    const tool = this.#context.getTool(name);
    if (!tool) return { ok: false, error: `Unknown tool: ${name}` };
    const schema = JSON.stringify(tool.inputSchema);
    const cached = this.#validators.get(name);
    const validate = cached?.schema === schema ? cached.validate : this.#ajv.compile(tool.inputSchema);
    this.#validators.set(name, { schema, validate });
    if (!validate?.(input)) {
      return {
        ok: false,
        error: "Invalid tool arguments",
        issues: (validate?.errors ?? []).map((error) => `${error.instancePath || "/"} ${error.message}`),
      };
    }
    try {
      const result = await tool.execute(input);
      this.#context.emit("tool/executed", { name, input, result });
      return result;
    } catch (error) {
      return { ok: false, error: error instanceof Error ? error.message : String(error) };
    }
  }
}

File: python_harness/agent.py
"""Chapter 1: the smallest useful agent loop."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Awaitable, Callable, Protocol


Json = dict[str, Any] | list[Any] | str | int | float | bool | None


class Llm(Protocol):
    async def complete(self, request: dict[str, Any]) -> dict[str, Any]: ...


ToolHandler = Callable[[dict[str, Any]], Awaitable[Json]]


@dataclass(frozen=True)
class Tool:
    name: str
    description: str
    parameters: dict[str, Any]
    execute: ToolHandler


def validate_arguments(schema: dict[str, Any], arguments: dict[str, Any]) -> str | None:
    """Validate the small JSON-Schema subset used by this tutorial."""
    required = schema.get("required", [])
    missing = [name for name in required if name not in arguments]
    if missing:
        return f"missing required fields: {', '.join(missing)}"
    properties = schema.get("properties", {})
    if schema.get("additionalProperties") is False:
        unknown = [name for name in arguments if name not in properties]
        if unknown:
            return f"unknown fields: {', '.join(unknown)}"
    return None


class Agent:
    def __init__(self, llm: Llm, tools: list[Tool], *, max_steps: int = 8) -> None:
        self.llm = llm
        self.tools = {tool.name: tool for tool in tools}
        self.max_steps = max_steps

    async def run(self, objective: str) -> str:
        messages: list[dict[str, Any]] = [{"role": "user", "content": objective}]
        schemas = [
            {"name": tool.name, "description": tool.description, "parameters": tool.parameters}
            for tool in self.tools.values()
        ]

        for step in range(1, self.max_steps + 1):
            request = {
                "system": "使用给定工具完成目标；根据工具结果继续，完成后停止调用工具。",
                "tools": schemas,
                "messages": messages,
                "dynamic_context": f"step={step}",
            }
            response = await self.llm.complete(request)
            messages.append({"role": "assistant", "content": response.get("content", "")})
            calls = response.get("tool_calls", [])
            if not calls:
                return str(response.get("content", ""))

            for call in calls:
                name, arguments = call["name"], call.get("arguments", {})
                tool = self.tools.get(name)
                error = "unknown tool" if tool is None else validate_arguments(tool.parameters, arguments)
                result: Json = {"error": error}
                if tool is not None and error is None:
                    try:
                        result = await tool.execute(arguments)
                    except Exception as exc:  # Tool errors become model-readable results.
                        result = {"error": str(exc)}
                messages.append({"role": "tool", "name": name, "content": result})

        raise RuntimeError(f"agent exceeded {self.max_steps} steps")
